### PR TITLE
Transactions logging and lifecycle

### PR DESCRIPTION
**Pull request by hackaugusto**
## Description

~~Review/Merge after #5879~~

This PR changes how transactions are sent, there are few improvements:

- explicit lifecycle using different types
- better logging
- less duplicate code (for the smart contract calls)

### CODE REVIEW

File: raiden/network/proxies/secret_registry.py [hunk 1]
```diff
@@ -263,12 +253,11 @@ def _register_secret_batch(
             #
             # Either of these is a bug. The contract does not use
             # assert/revert, and the account should always be funded
-            assert gas_limit
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: I believe this was a bug.

File: raiden/network/proxies/secret_registry.py [hunk 2]
```diff
@@ -134,43 +131,34 @@ def register_secret_batch(self, secrets: List[Secret]) -> None:
         # From here on the lock is not required. Context-switches will happen
         # for the gas estimation and the transaction, however the
         # synchronization data is limited to the open_secret_transactions
-        log_details = {
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: the `node`, `contract`, and function arguments are now logged by default. Just kept the `secrethashes_not_sent` for additional context.

File: raiden/network/proxies/secret_registry.py [hunk 3]
```diff
@@ -134,43 +131,34 @@ def register_secret_batch(self, secrets: List[Secret]) -> None:
...
 
-        if gas_limit:
-            gas_limit = safe_gas_limit(
-                gas_limit, len(secrets_to_register) * GAS_REQUIRED_PER_SECRET_IN_BATCH
+        if estimated_transaction is not None:
+            estimated_transaction.estimated_gas = safe_gas_limit(
+                estimated_transaction.estimated_gas,
+                len(secrets_to_register) * GAS_REQUIRED_PER_SECRET_IN_BATCH,
             )
-            log_details["gas_limit"] = gas_limit
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: gas limit is also included by default

File: raiden/network/proxies/secret_registry.py [hunk 4]
```diff
@@ -134,43 +131,34 @@ def register_secret_batch(self, secrets: List[Secret]) -> None:
...
 
             try:
-                transaction_hash = self.client.transact(
-                    self.proxy, "registerSecretBatch", gas_limit, secrets_to_register
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: the arguments don't have to be duplicated anymore, they are part of the `transaction` object, created by `self.client.estimate_gas`

File: raiden/network/proxies/service_registry.py [hunk 5]
```diff
@@ -140,16 +137,17 @@ def set_url(self, url: str) -> None:
             msg = "URL provided to service registry must be a valid HTTP(S) endpoint."
             raise BrokenPreconditionError(msg)
 
-        with log_transaction(log, "set_url", log_details):
```
Comment (hackaugusto) [inline raiden/network/proxies/service_registry.py]: the context manager is not necessary anymore, logging is done by the `transaction` objects.

File: raiden/network/rpc/client.py [hunk 6]
```diff
@@ -443,88 +481,164 @@ def monkey_patch_web3(web3: Web3, gas_price_strategy: Callable) -> None:
     Eth.call = patched_web3_eth_call
 
 
-class TransactionSlot:
```
Comment (hackaugusto) [inline raiden/network/rpc/client.py]: I'm forcing the creation of the slot to go through the client by hiding the implementation. This prevents using nonces wrongly.

File: raiden/network/rpc/client.py [hunk 7]
```diff
@@ -443,88 +481,164 @@ def monkey_patch_web3(web3: Web3, gas_price_strategy: Callable) -> None:
...
+            node_gas_price=self.data.contract.web3.eth.gasPrice,
+        )
+
+        return transaction_estimated
+
+
+@dataclass
+class TransactionEstimated:
+    from_address: Address
+    data: Union[SmartContractCall, ByteCode]
```
Comment (hackaugusto) [inline raiden/network/rpc/client.py]: A eth transfer has a constant gas usage, that is why it is not in this union.

File: raiden/tests/integration/network/proxies/test_secret_registry.py [hunk 8]
```diff
@@ -149,19 +149,15 @@ def test_concurrent_secret_registration(secret_registry_proxy: SecretRegistry, m
         count: Dict[Secret, int] = defaultdict(int)
         transact = secret_registry_proxy.client.transact
 
-        # Using positional arguments with the signature used by SecretRegistry
-        # to call the Contract in order to have access to the secret.
-        # Monkey patching the proxy function because the test must registered
-        # the data that is effectively sent with the transaction.
-        def count_transactions(proxy, function_name, startgas, secrets):
-            for secret in secrets:
+        def count_how_many_times_a_secret_is_sent(transaction):
```
Comment (palango) [inline raiden/tests/integration/network/proxies/test_secret_registry.py]: A type here would help

File: raiden/network/proxies/service_registry.py [hunk 9]
```diff
@@ -140,16 +137,17 @@ def set_url(self, url: str) -> None:
...
-                raise RaidenUnrecoverableError(msg)
-
-            log_details["gas_limit"] = gas_limit
-            transaction_hash = self.client.transact(self.proxy, "setURL", gas_limit, url)
-            receipt = self.client.poll_transaction(transaction_hash)
-            failed_receipt = check_transaction_threw(receipt=receipt)
-            if failed_receipt:
-                msg = f"URL {url} is invalid"
-                raise RaidenUnrecoverableError(msg)
+        extra_log_details: Dict[str, Any] = {}
```
Comment (palango) [inline raiden/network/proxies/service_registry.py]: Why even instantiate this, seems to be unused other than next line?

File: raiden/network/proxies/token.py [hunk 10]
```diff
@@ -78,74 +77,31 @@ def approve(self, allowed_address: Address, allowance: TokenAmount) -> None:
...
-                "allowance": allowance,
-            }
-
-            with log_transaction(log, "approve", log_details):
-                checking_block = self.client.get_checking_block()
-                error_prefix = "Call to approve will fail"
-                gas_limit = self.client.estimate_gas(
-                    self.proxy, checking_block, "approve", allowed_address, allowance
-                )
+            log_details: Dict[str, Any] = {}
```
Comment (palango) [inline raiden/network/proxies/token.py]: Same here

File: raiden/network/proxies/token.py [hunk 11]
```diff
@@ -203,68 +191,57 @@ def transfer(self, to_address: Address, amount: TokenAmount) -> None:
...
-                "amount": amount,
-            }
-
-            with log_transaction(log, "transfer", log_details):
-                checking_block = self.client.get_checking_block()
-                gas_limit = self.client.estimate_gas(
-                    self.proxy, checking_block, "transfer", to_address, amount
-                )
-                failed_receipt = None
+            log_details: Dict[str, Any] = {}
```
Comment (palango) [inline raiden/network/proxies/token.py]: Same here

File: raiden/network/proxies/service_registry.py (hunk 9 above)
Comment (hackaugusto) [reply raiden/network/proxies/service_registry.py]: typing

**Pull request merged**